### ISSUE TITLE
The « record » variable could be called alone, but currently you can only query it with row.record

### ISSUE THREAD

**Issue by antoine2711** [OpenRefine/OpenRefine#2523]
**Describe the bug**
In the Transform dialog, you can enter variables. The documentation is not clear about which can be called directly, and which can only be called thru another variable.

Since, in the logic, rows are enclosed by record, intuitively, someone would expect this to exist: `records.rows`. It does, but only by `row.record.rows`. It's not natural, longer, and harder to read.

![image](https://user-images.githubusercontent.com/12217525/78605826-3ddbdc80-782a-11ea-9eb4-815ceb4a0c5f.png)

**To Reproduce**
1. Go to 'Cell -> Transform'
2. Type 'record'
4. See `null` in the preview

calculation|now|wished
-|-|-
row.record|[object WrappedRecord]|[object WrappedRecord]
record|null|[object WrappedRecord]
cell.recon|[object Recon]|[object Recon]
recon|null|[object Recon]

**Current Results**
`null` in the preview

**Expected behavior**
To see, just as if I've entered `row.record`, `[object WrappedRecord]`.

**OpenRefine (please complete the following information):**
 - Version 3.3

**Additional details**
This issue and request is trivial. But, it's also very simple to code. It will also make easier the variable model of OR, and, be logical with the ideal of expanding the metadata usage in OR. Also, the syntax of GREL being by nature harder to read and write, small changes like this help to palliate this weakness of OR. It's no big improvement, but it doesn't break anything, doesn't go against the philosophy of OR and is simple to implement and test.

Regards, Antoine

**Comment by thadguidry** [comment]
They are called subfields.
@wetneb and I already had a big discussion about this and improving the docs would be useful.
I'd make that table into more than just 2 columns so that folks can actually see the subfields for what they really are.

Variable Name | Subfield | Meaning | Example
------------ | ---- | ------------- | --------
row | | The current row; ... | 
" | record | one or more rows grouped... | row.record()
  
Furthermore, we do indeed plan to add more hinting and GREL syntax helpers.  We have discussed this many times in the past on various issues to improve the expression preview ui.
Go Crazy: https://github.com/OpenRefine/OpenRefine/labels/expression%20preview%20ui

**Comment by antoine2711** [comment]
> They are called subfields.

Yes, I understand that. Just like `recon.match` is also. I understand that this could go far. (I use a lot `cell.recon.match.id`, and having `match.id` could be simpler. I'd admit it' I'm also tempted to ask for a `match` global variable that would point to `cell.recon.match`…)

But `record`, is something one would expect, much more than `row.record`.
I'll admit, `match` -> `cell.recon.match` would be more useful than `recon` -> `cell.recon`. I guess I'm just opening a discussion here, because I didn't found it anywhere else.

And yes, I've read most of the « Expression Preview UI » issues, and most of the 9 are very interesting. Thanks for you input.

Regards, Antoine

**Comment by thadguidry** [comment]
@antoine2711 Sure.   Your original intent is captured well in this statement:
> The documentation is not clear about which can be called directly

So, I generally think a better approach instead of command aliases as you are suggesting we would improve the Expression Preview UI and offer some amount of code completion like IDEs do.

Imagine you type `match` and code completion overlays `cell.recon.match` just like your IDE would do.  Example: https://codemirror.net/demo/complete.html
As well as a help panel that has a search input that can be toggled and slides out on left or right, for example, that reacts to what users type in would help users more.  In general, the Expression Preview UI could definitely be made bigger optionally based on what the user toggles on/off for assistance.

**Comment by antoine2711** [comment]
@thadguidry: everything you said in your last comment, I share it. It would be much better to have code completion, a better help, color syntaxing than these command aliases. 

If you ask me what I prefer, I'd go for those features right away. But if you ask me what I want to code…

The way I see it, and I might be wrong, is that this issue is a matter of 3/4 to 1h for a normal programer. Code completion, on the other hand, probably around 30h for a GOOD programer.

What I propose, we could have in v3.4, but code completion before v4.1? Seams pretty optimistic. Also, after this issue (or while it stalls), I'd like to take a swing at `project.name`, `project.id`, `project.importations`, then, if successful, `cell.column`, `columns`, and so on…

I think metadatas could get some love…

Regards, Antoine

**Comment by allanaaa** [comment]
Thanks for posting this! I'm testing how some of these variables work for the documentation, and was tripped up on this for 5-10 minutes before I searched and found this issue. I definitely think `record` and `recon` should work on their own. 
In the meantime the documentation will use cell.recon and row.record everywhere.

**Comment by tfmorris** [comment]
I'm not sure I follow the discussion.

@thadguidry do you object to the proposed change of making `record` and `recon` top level variables? If so, why?

**Comment by thadguidry** [comment]
@tfmorris not objecting.  My stance will always be that what simplifies things for our users (if we do the hard parts) will be best.  This issue proposed simplifies things in my opinion.  I was just giving @antoine2711 context of what we do about documenting the current deeper subfields and what we could do currently to make it easier for users from a doc perspective (which is now resolved on the wiki).  Anyways, I agree with @antoine2711 proposal where it makes sense to have those deeper subfields for the parts that users know and that we have menu items for such as `record` and `recon`, to be brought up as primary objects in GREL.  I did +1 this issue proposed.

**Comment by tfmorris** [comment]
Due to our current setup/evaluation scheme there is some overhead to setting these up before every evaluation, but it doesn't affect any inner loops, so probably isn't a big deal.

Summary of the current state of play:

Row fields (* = visible at top level)
- index * (as rowIndex)
- record
- columnNames
- flagged
- starred

Cell fields (none visible at top level)
- recon
- error message

Other top level variables:
- value
- cell
- row
- rowIndex
- cells
- columnName
- project
- PI
- true
- false

**Comment by allanaaa** [comment]
I had no idea "project" was a variable! What is the syntax? I'd like to add it to the docs. I'm guessing it's project["project name"] just like cells["column name"]?

**Comment by ostephens** [comment]
I also didn't know "project" was a variable - and I'm not sure what it can actually be used for!

**Comment by tfmorris** [comment]
OK, I guess that was a little bit of a red herring. It *does* get passed in to the expression evaluation context, but it **doesn't** have any fields, making it not terribly useful.

Having said that, one thing that you **can** do is `project.jsonize()` to get its contents serialized as a string and you could presumably then parse the resulting JSON and play with it, but I'd recommend that we not document it. No sense introducing more API surface area that we need to keep stable.

> I'm guessing it's project["project name"] just like cells["column name"]?

No, that would be a different imaginary variable named `projects` or `workspace` or some such. The imaginary variable `project` would have fields like `id`, `metadata`, `name`, etc. You can see its contents if you use the hack above.

**Comment by allanaaa** [comment]
That's really interesting. I could see using some project metadata to, say, determine whether you want to perform a cross() from another project into your current project or not (based on a modification date or something?). But I guess not, if we only have a current project variables and not a projects variables.  Yes, it's probably worth hiding for now.

**Comment by wetneb** [comment]
Unless this is addressed soon I would say it is fine to leave this out of the upcoming 3.5 release. Feel free to add the milestone again if someone finds the time to submit a PR for it.

**Comment by tfmorris** [comment]
Also mentioned here: https://github.com/OpenRefine/OpenRefine/issues/3577#issuecomment-785094102
